From: Chmouel Boudjnah <chmouel@mandrakesoft.com>
Subject: [kernel] compile source with default config of running kernel
To: Mandrake Kernel <kernel@linux-mandrake.com>
Date: 08 Jun 2001 17:13:26 +0200


here is a patch (adapted from a va-kernel-rpm patch) to have
autoconfig when running a different kernel.

The idea is :

- we ship a kernel-source-2.4.5-1mdk to be compiled for UP kernel.
- but we are running a SMP kernel
- we detect what we are running with /boot/kernel.h (generated at boot)
- check if the /boot/config-2.4.5-1mdk$(detected smp or enterprise via/boot/kernel.h) not done when building rpm
- use it as default instead of arch/${ARCH}/defaultconfig if no .config 

Please Juan apply (tested)

Changelog:

- compile source with default config of running kernel.

--- linux/scripts/Configure.chmou	Fri Jun  8 15:50:54 2001
+++ linux/scripts/Configure	Fri Jun  8 17:07:04 2001
@@ -545,8 +545,16 @@
 fi
 
 DEFAULTS=arch/$ARCH/defconfig
+KVER=$(sed -n 's/#define UTS_RELEASE "\(.*mdk\).*"$/\1/p' include/linux/version.h)
+KTYPE=$(sed -n 's/^#define __BOOT_KERNEL_\([^ ]*\) 1/\1/p' /boot/kernel.h|tr A-Z a-z)
+[[ "$KTYPE" = "up" ]] && KTYPE=""
+KCONF=/boot/config-${KVER}${KTYPE}
+echo $KCONF
+
 if [ -f .config ]; then
   DEFAULTS=.config
+elif [[ -f "${KCONF}" && -z "${RPM_BUILD_ROOT}" ]];then
+  DEFAULTS="${KCONF}"
 fi
 
 if [ -f $DEFAULTS ]; then
--- linux/scripts/Menuconfig.chmou	Fri Jun  8 15:50:54 2001
+++ linux/scripts/Menuconfig	Fri Jun  8 17:05:34 2001
@@ -1389,8 +1389,15 @@
 fi
 
 DEFAULTS=arch/$ARCH/defconfig
+KVER=$(sed -n 's/#define UTS_RELEASE "\(.*mdk\).*"$/\1/p' include/linux/version.h)
+KTYPE=$(sed -n 's/^#define __BOOT_KERNEL_\([^ ]*\) 1/\1/p' /boot/kernel.h|tr A-Z a-z)
+[[ "$KTYPE" = "up" ]] && KTYPE=""
+KCONF=/boot/config-${KVER}${KTYPE}
+
 if [ -f .config ]; then
   DEFAULTS=.config
+elif [[ -f "${KCONF}" ]];then
+  DEFAULTS="${KCONF}"
 fi
 
 if [ -f $DEFAULTS ]
--- linux/scripts/Makefile.chmou	Fri Jun  8 15:50:57 2001
+++ linux/scripts/Makefile	Fri Jun  8 17:07:25 2001
@@ -16,7 +16,15 @@
 	fi
 	cat ${HEADER} >> ./kconfig.tk
 	./tkparse < ../arch/${ARCH}/config.in >> kconfig.tk
-	echo "set defaults \"arch/${ARCH}/defconfig\"" >> kconfig.tk
+	KVER=`sed -n 's/^.define UTS_RELEASE "\(.*mdk\).*"/\1/p' ../include/linux/version.h`; \
+	KTYPE=`sed -n 's/^.define __BOOT_KERNEL_\([^ ]*\) 1/\1/p' /boot/kernel.h|tr A-Z a-z`; \
+	[[ "$$KTYPE" = "up" ]] && KTYPE="" ;\
+	KCONF=/boot/config-$$KVER$$KTYPE ; \
+	if [ -f "$$KCONF" ]; then \
+		echo "set defaults \"$$KCONF\"" >> kconfig.tk; \
+	else \
+		echo "set defaults \"arch/${ARCH}/defconfig\"" >> kconfig.tk; \
+	fi
 	echo "set ARCH \"${ARCH}\"" >> kconfig.tk
 	cat ${TAIL} >> kconfig.tk
 	chmod 755 kconfig.tk
